drivers/metal/rendering_device_driver_metal.mm

```diff
@@ -0,0 +1,3919 @@
+				[enc copyFromTexture:texture
+									 sourceSlice:region.texture_subresources.base_layer + lyrIdx
+									 sourceLevel:mip_level
+									sourceOrigin:txt_origin
+									  sourceSize:txt_size
+										toBuffer:buffer
+							   destinationOffset:region.buffer_offset + (bytesPerImg * lyrIdx)
+						  destinationBytesPerRow:bytesPerRow
+						destinationBytesPerImage:bytesPerImg
+										 options:blit_options];
+			}
+		}
+	}
+}
+
+void RenderingDeviceDriverMetal::command_copy_buffer_to_texture(CommandBufferID p_cmd_buffer, BufferID p_src_buffer, TextureID p_dst_texture, TextureLayout p_dst_texture_layout, VectorView<BufferTextureCopyRegion> p_regions) {
+	_copy_texture_buffer(p_cmd_buffer, CopySource::Buffer, p_dst_texture, p_src_buffer, p_regions);
+}
+
+void RenderingDeviceDriverMetal::command_copy_texture_to_buffer(CommandBufferID p_cmd_buffer, TextureID p_src_texture, TextureLayout p_src_texture_layout, BufferID p_dst_buffer, VectorView<BufferTextureCopyRegion> p_regions) {
+	_copy_texture_buffer(p_cmd_buffer, CopySource::Texture, p_src_texture, p_dst_buffer, p_regions);
+}
+
+#pragma mark - Pipeline
+
+void RenderingDeviceDriverMetal::pipeline_free(PipelineID p_pipeline_id) {
+	MDPipeline *obj = (MDPipeline *)(p_pipeline_id.id);
+	delete obj;
+}
+
+// ----- BINDING -----
+
+void RenderingDeviceDriverMetal::command_bind_push_constants(CommandBufferID p_cmd_buffer, ShaderID p_shader, uint32_t p_dst_first_index, VectorView<uint32_t> p_data) {
+	MDCommandBuffer *cb = (MDCommandBuffer *)(p_cmd_buffer.id);
+	MDShader *shader = (MDShader *)(p_shader.id);
+	shader->encode_push_constant_data(p_data, cb);
+}
+
+// ----- CACHE -----
+
+String RenderingDeviceDriverMetal::_pipeline_get_cache_path() const {
+	String path = OS::get_singleton()->get_user_data_dir() + "/metal/pipelines";
+	path += "." + context_device.name.validate_filename().replace(" ", "_").to_lower();
+	if (Engine::get_singleton()->is_editor_hint()) {
+		path += ".editor";
+	}
+	path += ".cache";
+
+	return path;
+}
+
+bool RenderingDeviceDriverMetal::pipeline_cache_create(const Vector<uint8_t> &p_data) {
+	return false;
+	CharString path = _pipeline_get_cache_path().utf8();
+	NSString *nPath = [[NSString alloc] initWithBytesNoCopy:path.ptrw()
+													 length:path.length()
+												   encoding:NSUTF8StringEncoding
+											   freeWhenDone:NO];
+	MTLBinaryArchiveDescriptor *desc = [MTLBinaryArchiveDescriptor new];
+	if ([[NSFileManager defaultManager] fileExistsAtPath:nPath]) {
+		desc.url = [NSURL fileURLWithPath:nPath];
+	}
+	NSError *error = nil;
+	archive = [device newBinaryArchiveWithDescriptor:desc error:&error];
+	return true;
+}
+
+void RenderingDeviceDriverMetal::pipeline_cache_free() {
+	archive = nil;
+}
+
+size_t RenderingDeviceDriverMetal::pipeline_cache_query_size() {
+	return archive_count * 1024;
+}
+
+Vector<uint8_t> RenderingDeviceDriverMetal::pipeline_cache_serialize() {
+	if (!archive) {
+		return {};
```

Comment:
```suggestion
		return Vector<uint8_t>();
```

I know this is painful, but that's what the current codebase is using. We may want to have a discussion among Godot contributors to allow that in trivial cases.

Reply:
Fixed!